mithril-common/src/chain_reader/pallas_chain_reader.rs

```diff
@@ -0,0 +1,262 @@
+    ) -> StdResult<Option<ChainBlockNextAction>> {
+        let client = self.get_client().await?;
+        let chainsync = client.chainsync();
+
+        chainsync
+            .find_intersect(vec![point.to_owned().into()])
+            .await?;
+
+        let next = match chainsync.has_agency() {
+            true => chainsync.request_next().await?,
+            false => chainsync.recv_while_must_reply().await?,
+        };
+
+        self.process_next_chain_block(next).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use pallas_network::{
+        facades::NodeServer,
+        miniprotocols::{
+            chainsync::{BlockContent, Tip},
+            Point,
+        },
+    };
+    use tokio::net::UnixListener;
+
+    use super::*;
+
+    use crate::test_utils::TempDir;
+
+    /// Returns a fake chain point for testing purposes.
+    fn get_fake_chain_point_backwards() -> ChainPoint {
+        ChainPoint::from(Point::Specific(
+            1654413,
+            hex::decode("7de1f036df5a133ce68a82877d14354d0ba6de7625ab918e75f3e2ecb29771c2")
+                .unwrap(),
+        ))
+    }
+
+    /// Returns a fake chain point for testing purposes.
+    fn get_fake_chain_point_forwards() -> ChainPoint {
+        Tip(
+            Point::Specific(
+                1654413,
+                hex::decode("7de1f036df5a133ce68a82877d14354d0ba6de7625ab918e75f3e2ecb29771c2")
+                    .unwrap(),
+            ),
+            1337,
+        )
+        .into()
+    }
+
+    /// Returns a fake intersection point for testing purposes.
+    fn get_fake_intersection_point() -> Point {
+        Point::Specific(
+            1654413,
+            hex::decode("7de1f036df5a133ce68a82877d14354d0ba6de7625ab918e75f3e2ecb29771c2")
+                .unwrap(),
+        )
+    }
+
+    /// Creates a new work directory in the system's temporary folder.
+    fn create_temp_dir(folder_name: &str) -> PathBuf {
+        TempDir::create_with_short_path("pallas_chain_observer_test", folder_name)
+    }
+
+    /// Sets up a mock server for related tests.
+    ///
+    /// Use the `intersections` parameter to define exactly how many
+    /// local state queries should be intersepted by the `mock_server`
+    /// and avoid any panic errors.
+    async fn setup_server(socket_path: PathBuf, action: String) -> tokio::task::JoinHandle<()> {
+        tokio::spawn({
+            async move {
+                if socket_path.exists() {
+                    fs::remove_file(&socket_path).expect("Previous socket removal failed");
+                }
+
+                let known_point = get_fake_intersection_point();
+                let unix_listener = UnixListener::bind(socket_path.as_path()).unwrap();
+                let mut server = NodeServer::accept(&unix_listener, 10).await.unwrap();
+
+                let chansync_server = server.chainsync();
+
+                // chansync_server receives request next from client, sends rollbackwards
```

Comment:
This comment and the one below do not seem to be describing what's happening the code :thinking: